contracts-periphery/script/Deploy.s.sol

```diff
@@ -1,15 +1,35 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.16;
 
-import "forge-std/Script.sol";
-import {UmbraBatchSend} from "src/UmbraBatchSend.sol";
-import {IUmbra} from "src/interface/IUmbra.sol";
+/// @title Deploy a contract to multiple networks.
+/// @notice Run the script with `forge script Deploy --private-key <privateKey>`
+/// and add `--broadcast` to broadcast the tx.
+/// @dev The contract that's being deployed is specified in `BaseScript.sol`.
+/// Specify additional netowrk aliases in [rpc_enpoints] in `forge.toml`
+/// before adding them to the list of networks below.
+/// Add corresponding `env` variables in a `.env` file, such as ETH_RPC_URL=<URL>
 
-contract Deploy is Script {
-  address constant UMBRA = 0xFb2dc580Eed955B528407b4d36FfaFe3da685401;
+import {StdChains} from "forge-std/StdChains.sol";
+import {BaseScript} from "script/BaseScript.sol";
 
+contract Deploy is BaseScript {
+  /// The list of networks to deploy to.
+  string[] public networks = ["mainnet", "optimism", "arbitrum_one", "polygon", "goerli", "sepolia"];
+
+  /// @notice Deploy the contract to the list of networks,
+  /// Will run by default when executing the script.
   function run() public {
-    vm.broadcast();
-    new UmbraBatchSend(IUmbra(UMBRA));
+    /// Compute the address the contract will be deployed to
+    address expectedContractAddress = computeCreateAddress(msg.sender, vm.getNonce(msg.sender));
+
+    /// Turn off fallback to default RPC URLs since they can be flaky.
+    setFallbackToDefaultRpcUrls(false);
+
+    /// Check if the contract is already deployed on the network, if not, deploy it.
+    for (uint256 i; i < networks.length; i++) {
+      vm.createSelectFork(getChain(networks[i]).rpcUrl);
+      bool isDeployed = address(expectedContractAddress).code.length > 0;
+      if (!isDeployed) deploy();
+    }
```

Comment:
There's a few issues here:
1. Nonces differ by chain, but we check the nonce once upfront, before we select a fork, and use that for all chains.
2. With this logic, we'll always deploy to every network. What this does is compute the expected address for the `msg.sender` using their current nonce, which is the nonce used by the next tx. By definition, there will never be a contract there

We need something like this, which assumes the expected nonce we want to deploy at is 1. Le me know if you have any questions on this. I'd also suggest writing some tests for your script so you can verify it does what you expect :) 

```solidity
contract Deploy is BaseScript {
  // TODO Get a real value here. The nonce we need to deploy the contract on each chain.
  uint256 constant EXPECTED_NONCE = 1;

  // The list of networks to deploy to.
  string[] public networks = ["mainnet", "optimism", "arbitrum_one", "polygon", "goerli", "sepolia"];

  function run() public {
    address expectedContractAddress = computeCreateAddress(msg.sender, EXPECTED_NONCE);

    for (uint256 i; i < networks.length; i++) {
      vm.createSelectFork(getChain(networks[i]).rpcUrl);
      bool isDeployed = address(expectedContractAddress).code.length > 0;

      if (isDeployed) {
        console2.log("Skipping '%s': contract already deployed", networks[i]);
        continue;
      }

      uint256 nonce = vm.getNonce(msg.sender);
      if (nonce > EXPECTED_NONCE) {
        console2.log(
          "Skipping '%s': current nonce %d > expected nonce %d", networks[i], nonce, EXPECTED_NONCE
        );
        continue;
      }

      if (nonce < EXPECTED_NONCE) {
        console2.log(
          "Skipping '%s': current nonce %d < expected nonce %d", networks[i], nonce, EXPECTED_NONCE
        );
        continue;
      }

      deploy();
      console2.log("Deployed contract to '%s' at %s", networks[i], expectedContractAddress);
    }
  }
}
```

Reply:
I was making the assumption that the nonce will be the same across chains since we're trying to deploy the contract to have the same addresses. But even then, after `console2.log`ging I see how 1. can be an issue. And 2. was definitely a big miss.

My only question then for the smoothest DX, should we ask the deployer to manually edit the script file to enter the `EXPECTED_NONCE`, or should we make it an argument in `run()`, or is there a better way?

Reply:
Just hardcode IMO. Once it’s set it should never change, so let’s keep it simple. Once the value is set and shouldn’t be changed we can update the code comment to reflect that

Reply:
Cool. Updated here: https://github.com/ScopeLift/umbra-protocol/pull/504/commits/9bfbf40edf11a8894d7efadeb07a8b313fee229a